# Build a Logistic Regression Model with Scikit-learn?


In [28]:
import pandas as pd 

df = pd.read_csv("C:/Users/kepohin/OneDrive/titanic.csv")
df.head()

Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children     Fare
0         0       3    male  22.0                 1                 0   7.2500
1         1       1  female  38.0                 1                 0  71.2833
2         1       3  female  26.0                 0                 0   7.9250
3         1       1  female  35.0                 1                 0  53.1000
4         0       3    male  35.0                 0                 0   8.0500

# Prep Data with Pandas


In [47]:
df['male'] = df['Sex'] == 'male'
df.head()


Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children  \
0         0       3    male  22.0                 1                 0   
1         1       1  female  38.0                 1                 0   
2         1       3  female  26.0                 0                 0   
3         1       1  female  35.0                 1                 0   
4         0       3    male  35.0                 0                 0   

      Fare   male  
0   7.2500   True  
1  71.2833  False  
2   7.9250  False  
3  53.1000  False  
4   8.0500   True

In [35]:
# we need to make all our Features columns(Pclass', 'male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare') numerical: using numpy array
x = df[['Pclass', 'male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare']].values
print(x)

[[3 True 22.0 1 0 7.25]
 [1 False 38.0 1 0 71.2833]
 [3 False 26.0 0 0 7.925]
 ...
 [3 False 7.0 1 2 23.45]
 [1 True 26.0 0 0 30.0]
 [3 True 32.0 0 0 7.75]]


In [36]:
# Now let’s take the target (the Survived column) and store it in a variable y.
y = df['Survived'].values
print(y)

[0 1 1 1 0 0 0 0 1 1 1 1 0 0 0 1 0 1 0 1 0 1 1 1 0 1 0 0 1 0 0 1 1 0 0 0 1
 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 1 0 1 1 0 1 0 0 1 0 0 0 1 1 0 1 0 0 0 0 0 1
 0 0 0 1 1 0 1 1 0 1 1 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 1 1 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 1 0 0 0 0 1 0 0 1 0 0 0 0 1 1 0 0 0 1 0 0
 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0
 1 1 0 0 1 0 1 1 1 1 0 0 1 0 0 0 0 0 1 0 0 1 1 1 0 1 0 0 0 1 1 0 1 0 1 0 0
 0 1 0 1 0 0 0 1 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 1 1 0 0 0 0 0 0 1 1 1 1 1
 1 0 0 0 0 0 1 1 1 0 1 1 0 1 1 0 0 0 1 0 0 0 1 0 0 1 0 1 1 1 1 0 0 0 0 0 0
 1 1 1 1 0 1 0 1 1 1 0 1 1 1 0 0 0 1 1 0 1 1 0 0 1 1 0 1 0 1 1 1 1 0 0 0 1
 0 0 1 1 0 1 1 0 0 0 1 1 1 1 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 1 1 1 1 1 0
 0 0 0 1 1 0 0 0 1 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 0 0 1 1 0 0 0 0 0 0 1 0
 0 0 0 1 0 1 0 1 1 0 0 0 0 0 0 0 1 1 0 1 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 1 1
 1 1 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 0 0 0 0 1 0 0 1 1 0 0 0 0 0 1 0 0 0 1
 1 0 1 0 0 1 0 0 0 0 0 0 

# Build a Logistic Regression Model with Sklearn


In [41]:
# We start by importing the Logistic Regression model: all sklearn are built as classes

from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

In [42]:
x = df[['Age', 'Fare']].values
y = df['Survived'].values

In [43]:
model.fit(x, y)

LogisticRegression()

# Make Predictions with the Model


In [50]:
x = df[['Pclass', 'male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare']].values
y = df['Survived'].values

model = LogisticRegression()
model.fit(x, y)

LogisticRegression()

In [54]:
# Now we can use the predict method to make predictions.
model.predict(x)

array([0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,

In [55]:
df.head()

Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children  \
0         0       3    male  22.0                 1                 0   
1         1       1  female  38.0                 1                 0   
2         1       3  female  26.0                 0                 0   
3         1       1  female  35.0                 1                 0   
4         0       3    male  35.0                 0                 0   

      Fare   male  
0   7.2500   True  
1  71.2833  False  
2   7.9250  False  
3  53.1000  False  
4   8.0500   True

In [60]:
# 0
print(model.predict([[3, True, 22.0, 1, 0, 7.25]]))

[0]


In [61]:
# 1
print(model.predict([[1, False, 38.0, 1, 0, 71.28]]))

[1]


Let’s see what the model predicts for the first 5 rows of data and compare it to our target array. We get the first 5 rows of data with X[:5] and the first 5 values of the target with y[:5].

In [64]:
print(model.predict(x[:5]))

[0 1 1 1 0]


# Score the Model



#Score the Model
We can get a sense of how good our model is by counting the number of datapoints it predicts correctly. This is called the accuracy score.

In [65]:
# Let’s create an array that has the predicted y values.

y_pred = model.predict(x)

# Now we create an array of boolean values of whether or not our model predicted each passenger correctly.
y == y_pred


In [67]:
# To get the number of these that are true, we can use the numpy sum method.
print((y == y_pred).sum())

714


This means that of the 887 datapoints, the model makes the correct prediction for 714 of them.


In [68]:
# To get the percent correct, we divide this by the total number of passengers. We get the total number of passengers using the shape attribute.
y.shape[0]

887

In [72]:
print((y == y_pred).sum() / y.shape[0])


0.8049605411499436


# Score the Model: Exercise
Assume that y=[0, 0, 0, 1, 1] and the result of model.predict(X) is [0, 0, 1, 1, 0]. What is the expected output of the following code?

model.score(X, y)

In [76]:
from sklearn.metrics import accuracy_score

# True class
y = [0, 0, 0, 1, 1] 
# Predicted class
y_hat = [0, 0, 1, 1, 0]
accuracy_score(y, y_hat)

0.6